hb_get_preview3

jstebbins opened the pull request:
**Description of Change:**
New API that permits application of HB filters to preview images

**Test on:**

- [ ] Windows 10+  (via MinGW)
- [ ] macOS 10.13+
- [x] Ubuntu Linux

jstebbins committed:
muxavformat: fix compiler indentation warning

jstebbins committed:
Add ability to apply filters to previews

New APIs hb_get_preview3 and JSON version hb_get_preview3_json to
retrieve a cached preview image and apply relevant filters from an
hb_job_t to the image.

Returned image also has PAR applied, i.e. PAR of image is 1:1

jstebbins committed:
LinGui: use new api hb_get_preview3 for previews

Enables display of effects of filters on preview images

sr55 commented:
![image](https://user-images.githubusercontent.com/628593/113626018-87d3ea00-9659-11eb-8e5d-4b7813695716.png)

Confirm this works nicely on windows. 
- Padding
- Rotation
- Crop
- Dimensions
All seem to behave fine in testing. 


I have triggered a crash condition though. Let me try nail this down

sr55 commented:
Got some image corruption. Rotate 90, then pad:

![image](https://user-images.githubusercontent.com/628593/113627260-2745ac80-965b-11eb-8ca4-552d454cfd27.png)

![image](https://user-images.githubusercontent.com/628593/113627299-3593c880-965b-11eb-9916-e50e5c0417e7.png)

jstebbins committed:
pad: prevent invalid settings reaching avfilter

avfilter does not gracefully handle invalid pad width or height

jstebbins commented:
The image corruption is when pad fails to initialize due to invalid settings.  width must be >= input width and height must be >= input height. I've added sanitizing code to pad.c to prevent that.

galad87 commented:
It returns a 0x0 image when the colourspace filter is on.
Everything else seems to work.

jstebbins commented:
@galad87 thanks, I don't have colourspace filter in lingui, but I can fake it with a bit of code. I'll have a look.

jstebbins commented:
@galad87 can you tell me what values you are using for the filter. I can only seem to reproduce with invalid settings.

jstebbins commented:
It would help if you could add a log in hbavfilter.c to show the avfilter graph string that is being configured, just before avfilter_graph_parse2 around line 108.

galad87 commented:
`(char *) settings_str = 0x00007feed4441d70 "scale='width=1920:height=1080:flags=lanczos+accurate_rnd:in_color_matrix=bt709:out_color_matrix=bt709:out_range=limited',format='pix_fmts=yuv420p',zscale='primaries=bt470bg:transfer=709:matrix=smpte170m',format='pix_fmts=yuv420p',scale='width=1920:height=1080:flags=lanczos+accurate_rnd:in_color_matrix=smpte170m:out_color_matrix=smpte170m:out_range=limited',format='pix_fmts=bgra'"`

jstebbins commented:
Strange, with exactly the same settings, it is working for me.  Are you seeing something like "no path between colorspaces" in your log?  That's what I saw with invalid settings. Any other error messages in the log?

galad87 commented:
Setting the color tag on the in buffer fixes it: https://gist.github.com/galad87/eb59d481c76af6b8f4a47aa4ecf56497
But there should be a way to get the actual color tag of the outputted image.

jstebbins commented:
Ah, that kind of makes sense, although I don't see why we get different results.  AVCOL_SPC_RGB == 0. Those values should always be 0 in a newly created hb_buffer_t.  Normally they would get filled in from an AVFrame.  But in this case that information isn't stored in the preview file, so we need to fill in from the title.

jstebbins committed:
rotate: invert par when rotating by 90

How this was missed for so long, I don't know.  I guess nobody ever used
anamorphic settings with rotate.

jstebbins committed:
preset: don't add rotate to job when "angle=0:hflip=0"

jstebbins commented:
I think we've wrung out most of the issues. If everyone is ok with this, I'll clean up the commit history, force push, then merge.

galad87 commented:
Ship it!

sr55 commented:
ok, I'll push the WinUI shortly after you put this in.
